### ISSUE TITLE
Very slow indexing

### ISSUE THREAD

**Issue by Loremaster** [ankane/searchkick#334]
Hi. 

We use your gem and it indexes our data very slowly. Our log looks like this: 

```
** [out :: unicorn@10.20.30.16] BuildingMaterial Import (13524.1ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (15669.0ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (15291.3ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (15685.4ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (13987.4ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (13442.6ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (17599.4ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (15579.6ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (14367.3ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (14824.6ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (15662.3ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (16813.1ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (15924.0ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (14052.1ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (14864.0ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (15285.1ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (20349.8ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (21749.4ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (23592.1ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (18884.5ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (21063.3ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (21843.3ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (21773.6ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] BuildingMaterial Import (14658.5ms)  {"count":600}
 ** [out :: unicorn@10.20.30.16] Reindexing Machine...
 ** [out :: unicorn@10.20.30.16] Machine Import (108406.4ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (95769.8ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (83516.0ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (72669.3ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (56985.6ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (59733.3ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (57863.4ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (58632.4ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (73757.4ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (78712.6ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (60410.1ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (60231.2ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (60861.8ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (71683.7ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (71078.1ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Machine Import (17378.6ms)  {"count":270}
 ** [out :: unicorn@10.20.30.16] Reindexing Project...
 ** [out :: unicorn@10.20.30.16] Project Import (24.9ms)  {"count":11}
 ** [out :: unicorn@10.20.30.16] Reindexing Resume...
 ** [out :: unicorn@10.20.30.16] Resume Import (17.9ms)  {"count":2}
 ** [out :: unicorn@10.20.30.16] Reindexing Service...
 ** [out :: unicorn@10.20.30.16] 
 ** [out :: unicorn@10.20.30.16] Service Import (86577.2ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Service Import (87083.8ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Service Import (89190.9ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Service Import (88896.6ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Service Import (96774.8ms)  {"count":1000}
 ** [out :: unicorn@10.20.30.16] Service Import (39652.6ms)  {"count":457}
```

As you can see we have few heavy models, but they don't have so much data in them. Here is example of index for one of the model: 

``` ruby
class BuildingMaterial < ActiveRecord::Base
  # ...

  def search_data
    {
      title: title,
      material_brand_name: try(:material_brand).try(:name),
      material_model_name: try(:material_model).try(:name),
      tag_list: tags.pluck(:name),

      # Where
      building_material_category_id: building_material_category_id,
      material_brand_id: material_brand_id,
      material_model_id: material_model_id,
      country_code: country_code,
      shipping_types: shipping_types,
      created_at: created_at,
      search_vat_price: search_vat_price,

      city_latitude: city_latitude,
      city_longitude: city_longitude,
      region_latitude: region_latitude,
      region_longitude: region_longitude,
      country_latitude: country_latitude,
      country_longitude: country_longitude,
    }
  end
end
```

Some of this fields are strings, some of them are integers, some are floats. The full reindex takes near 33 minutes and that is for ~ 20000 records. It's very, very slow. Is there any way to improve speed of the indexing? Probably parallel it? What can I do in such situation?

**Comment by Chocksy** [comment]
I've created a gist with what @ankane said above. It seems to be working for me.
```ruby
namespace :searchkick do
  desc "maybe reindex without running out of memory?"
  task :altindex, [:start_index, :batch_size, :new_index_name] => :environment do |t, args|
    ActiveRecord::Base.logger = Logger.new(STDOUT)
    count = Item.count
    start = args[:start_index].to_i || 0
    batch_size = args[:batch_size].to_i || 1000

    # we create a new index or take one already present
    if args[:new_index_name]
      new_index = Searchkick::Index.new(args[:new_index_name], options)
      puts "we took a present index"
    else
      new_index = Item.searchkick_index.create_index
      puts "the new created index name is: #{new_index.name}"
    end

    puts "Starting reindex at #{start} with batch size of #{batch_size}..."
    Item.find_in_batches(:start=> start, :batch_size=> batch_size).with_index do |batch, index|
      puts "Processing batch ##{index}; total #{index*batch_size+batch_size} items"
      before = Time.now
      new_index.import(batch)
      time = Time.now - before
      puts "Processed #{index*batch_size} items in #{time} secs"
    end
  end
end
```
https://gist.github.com/Chocksy/6efa4833989f51e46fca

I have an issue with carrierwave though. I was able before to run the searchkick default rake task for my 1 mil records but now when i call the column of carrierwave after i moved to active_record it takes too long to get the path of the photo and the script breaks.

I'll try to find a solution and maybe make a wiki to help other people.

**Comment by netwire88** [comment]
Mongoid doesn't have `find_in_batches` so not sure if the above Gist would work for Mongoid.

**Comment by ankane** [comment]
Cleaning up issues

**Comment by MrHubble** [comment]
@ankane can you please confirm if the basic flow [you stated above](https://github.com/ankane/searchkick/issues/334#issuecomment-65186188) is still recommended for large datasets? 

Searchkick uses find_in_batches to by default to import documents. Therefore in the readme where it states:

>Reindex large set of records in batches
````
Product.where("id > 100000").find_in_batches do |batch|
  Product.searchkick_index.import(batch)
end
````

I think this example is only relevant where you want to reindex from a certain `id` onwards, not just reindexing large sets of records in batches, as I think that is done by default.

Also, maybe `resume` reduces the need for the basic flow:
>If a reindex is interrupted, you can resume it with:
`Product.reindex(resume: true)`

In summary, from what I have read, Searchkick `reindex` uses the bulk import API and `find_in_batches` by default and there is no reason to use the [steps discussed above](https://github.com/ankane/searchkick/issues/334#issuecomment-65186188). Is this correct?

**Comment by ankane** [comment]
With very large data sets, you may want to parallelize and/or have automatic retries when batches fail, so the above approach is still recommended if `Model.reindex` isn't working well for you. That being said, we should probably add automatic retries to the `reindex` method.

**Comment by MrHubble** [comment]
>you may want to parallelize

@ankane any tips on how to parallelize? My understanding is with zero downtime deployment we create a new index which we populate first before swapping to it. I'm following the gist by Chocksy [above](https://github.com/ankane/searchkick/issues/334#issuecomment-68989109) but I can't see how we could have zero downtime with 2 or more new indicies being created at the same time. By this I mean, in one task we create `products_index_1` and import 500 documents, and at the same time in another task we create `products_index_2` and import the next 500 documents. How would we then combine these into `new_combined_products_index`?
